src/common/token.ts

```diff
@@ -190,3 +203,145 @@ export class ArweaveToken implements TokenTools {
 export const WinstonToTokenAmount = (winston: BigNumber.Value) => winston;
 export const ARToTokenAmount = (ar: BigNumber.Value) =>
   new BigNumber(ar).times(1e12).valueOf();
+
+export class SolanaToken implements TokenTools {
+  protected logger: TurboLogger;
+  protected connection: Connection;
+  protected gatewayUrl: string;
+  protected pollingOptions: TokenPollingOptions;
+
+  constructor({
+    logger = new TurboWinstonLogger(),
+    gatewayUrl = 'https://api.mainnet-beta.solana.com',
+    pollingOptions = {
+      maxAttempts: 10,
+      pollingIntervalMs: 5_000,
+      initialBackoffMs: 7_000,
+    },
+  }: TokenConfig = {}) {
+    this.logger = logger;
+
+    this.gatewayUrl = gatewayUrl;
+    this.connection = new Connection(gatewayUrl, 'confirmed');
+    this.pollingOptions = pollingOptions;
+  }
+
+  public async createAndSubmitTx({
+    target,
+    tokenAmount,
+    signer,
+  }: TokenCreateTxParams): Promise<{
+    id: string;
+    target: string;
+  }> {
+    const publicKey = new PublicKey(bs58.encode(await signer.getPublicKey()));
+    const tx = new Transaction({
+      feePayer: publicKey,
+      ...(await this.connection.getLatestBlockhash()),
+    });
+
+    tx.add(
+      SystemProgram.transfer({
+        fromPubkey: publicKey,
+        toPubkey: new PublicKey(target),
+        lamports: +new BigNumber(tokenAmount),
+      }),
+    );
+
+    const serializedTx = tx.serializeMessage();
+    const signature = await signer.signData(serializedTx);
+
+    tx.addSignature(publicKey, Buffer.from(signature));
+
+    const id = bs58.encode(signature);
+    await this.submitTx(tx, id);
+
+    return { id, target };
+  }
+
+  private async submitTx(tx: Transaction, id: string): Promise<void> {
+    this.logger.debug('Submitting fund transaction...', { id });
+
+    await this.connection.sendRawTransaction(tx.serialize(), {
+      maxRetries: this.pollingOptions.maxAttempts,
+    });
+
+    await this.connection.confirmTransaction(
+      {
+        signature: id,
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        blockhash: tx.recentBlockhash!,
```

Comment:
can we just do a check above if it does not exist to avoid this null assertion? same as below?

Reply:
929bd19